File: test/bench/test_teleportation.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/Error.hpp"
#include "bench/Evaluation.hpp"
#include "bench/Teleportation.hpp"
#include "bench/TestUtils.hpp"

#include "gtest/gtest.h"

namespace test = mqt::bench::test;

namespace {

using mqt::bench::Output;
using mqt::bench::Teleportation;

TEST(Teleportation, ChecksTheTeleportedState) {
  const Teleportation benchmark;
  EXPECT_EQ(benchmark.output(), (Output{"result", 1}));
  EXPECT_DOUBLE_EQ(test::value(benchmark.probability("0")), 1.);
  EXPECT_DOUBLE_EQ(test::value(benchmark.probability("1")), 0.);
  EXPECT_EQ(test::errorKind(benchmark.probability("00")),
            mqt::bench::Error::Kind::InvalidArgument);
  EXPECT_EQ(test::errorKind(benchmark.probability("x")),
            mqt::bench::Error::Kind::InvalidArgument);

  const auto exact = test::value(benchmark.evaluate({{"0", 8}}));
  EXPECT_DOUBLE_EQ(exact.totalVariationDistance, 0.);
  EXPECT_DOUBLE_EQ(exact.squaredHellingerFidelity, 1.);
  EXPECT_EQ(exact.successProbability, 1.);

  const auto noisy = test::value(benchmark.evaluate({{"0", 6}, {"1", 2}}));
  EXPECT_DOUBLE_EQ(noisy.totalVariationDistance, 0.25);
  EXPECT_DOUBLE_EQ(noisy.squaredHellingerFidelity, 0.75);
  EXPECT_EQ(noisy.successProbability, 0.75);
}

} // namespace

File: test/python/bench/test_teleportation.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the teleportation benchmark."""

from __future__ import annotations

import json

import pytest

from mqt.core.bench import teleportation

from .utils import assert_generates


def test_teleportation_reference_json_and_generation() -> None:
    """Expose the fixed quantum teleportation benchmark without options."""
    benchmark = teleportation.Teleportation()
    assert benchmark.output.name == "result"
    assert benchmark.output.width == 1
    assert benchmark.probability("0") == 1
    assert benchmark.probability("1") == 0

    evaluation = benchmark.evaluate({"0": 128})
    assert evaluation.total_variation_distance == pytest.approx(0)
    assert evaluation.squared_hellinger_fidelity == pytest.approx(1)
    assert evaluation.success_probability == 1
    assert json.loads(benchmark.instance_specification_json)["parameters"] == {}

    instance_copy = teleportation.Teleportation.from_instance_specification_json(benchmark.instance_specification_json)
    manifest_copy = teleportation.Teleportation.from_manifest_json(benchmark.manifest_json)
    assert instance_copy.case_id == manifest_copy.case_id == benchmark.case_id

    assert_generates(benchmark.generate())
